### PR TITLE
feat(view): create a terminal output of the contents of the world

### PR DESCRIPTION
**Pull request by AtomToast**

### CODE REVIEW

File: src/view/terminalview.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+#ifndef EVE_TERMINALVIEW_H
+#define EVE_TERMINALVIEW_H
+
+#include "../model/world/world.h"
+
+class TerminalView {
+private:
+    std::string formatNumberSpacing(int number, int maxLen);
+public:
+    void render(World world);
```
Comment (anarchuser) [inline src/view/terminalview.h]: change all parameters to (const Class & class)
-> e.g. render(const World & world)

Comment (anarchuser) [reply src/view/terminalview.h]: if you only say (Class class) it would copy the whole object, causing potential performance issues.
(Class & class) gives you the reference, i.e. the object itself which is performant, but a security issue (since you could change the underlying object)

by specifying the parameter as (const Class & class) you get a constant reference, i.e. a view of the object which you can't change

File: test/view/terminalview.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        "ID:  24 | X:  138 | Y:  153 | Energy:  70\n"
+        "ID:  69 | X:  420 | Y:  420 | Energy:   5\n";
+
+    World world = World(1000, 1000, 40);
+    SECTION(""){}
+
+    Food food1 = Food(500, 689, 10);
+    SECTION("stringifyFood can convert a Food object into a string") {
+        REQUIRE(
+            terminalView.stringifyFood(food1) == "X:  500 | Y:  689 | Value:  10\n"
```
Comment (anarchuser) [inline test/view/terminalview.cpp]: the .stringifyFood method should be private and be called from within the .render method only.
that means you can't test it, but makes your object more clean and secure

File: src/view/terminalview.h [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    std::string createTable(World world);
+    std::string stringifyFood(Food food);
+    std::string stringifyEntity(Entity & entity);
```
Comment (anarchuser) [inline src/view/terminalview.h]: these three  method should be private and be called from within the .render method only.
that means you can't test it, but makes your object more clean and secure

File: src/view/terminalview.cpp [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    formattedXPos = formatNumberSpacing(xPos, 4);
+    formattedYPos = formatNumberSpacing(yPos, 4);
+    formattedValue = formatNumberSpacing(value, 3);
+
+    std::string outputString;
+    outputString += "X: " + formattedXPos + " | ";
+    outputString += "Y: " + formattedYPos + " | ";
+    outputString += "Value: " + formattedValue + "\n";
+    return outputString;
```
Comment (anarchuser) [inline src/view/terminalview.cpp:59]: you could make use of c++'s beautiful iostreaming:
i.e.:
return "X: " << formattedXPos << " | "
      << "Y: " << formattedYPos << " | "
      << "Value: " << formattedValue << "\n";

Comment (anarchuser) [reply src/view/terminalview.cpp:59]: I'm not 100% sure, but I think it's more performant than doing several independent string concatenations

File: src/view/terminalview.h (hunk 1 above)
Comment (AtomToast) [reply src/view/terminalview.h]: Does this also apply to the integers of the formatNumberSpacing function?

File: src/view/terminalview.h [hunk 5]
```diff
@@ -8,10 +8,10 @@ class TerminalView {
 private:
     std::string formatNumberSpacing(int number, int maxLen);
 public:
-    void render(World world);
-    std::string createTable(World world);
-    std::string stringifyFood(Food food);
-    std::string stringifyEntity(Entity & entity);
+    void render(const World world);
+    std::string createTable(const World world);
+    std::string stringifyFood(const Food food);
```
Comment (anarchuser) [inline src/view/terminalview.h]: Don't forget the ampersand &!
> (const World & world)
it's important

File: src/view/terminalview.h (hunk 3 above)
Comment (AtomToast) [reply src/view/terminalview.h]: I can understand the relevance of this but I encapsulating everything would just make working properly impossible

File: src/view/terminalview.cpp (hunk 4 above)
Comment (AtomToast) [reply src/view/terminalview.cpp:59]: Sounds cool, I'll look into it

File: src/view/terminalview.h (hunk 3 above)
Comment (anarchuser) [reply src/view/terminalview.h]: Those three are pure functions, right?
They only convert an object into a string (that's why it's called .stringify) but change nothing else.
That means there shouldn't be any reason to call them outside of the View class.
Following that, you can make them private.

Comment (anarchuser) [reply src/view/terminalview.h]: Just always ask yourself, in this order:
Does this need to be called from outside?
-> public

Does this need to be called from a subclass?
-> protected

Does this only need to be called.from within itself?
-> private

File: src/view/terminalview.cpp (hunk 4 above)
Comment (AtomToast) [reply src/view/terminalview.cpp:59]: I did some research and turns out it's best to use either += or .append() (they are pretty much the same under the hood)
Always avoid + since it creates a new array everytime.
<< should also be avoided unless you want to use to to concatenate with numbers it seems.
strcat() is outdated and slow

Look here for a quick reference https://www.thecodingdelight.com/string-cplusplus/

I will also write a research notion thing later

Comment (anarchuser) [reply src/view/terminalview.cpp:59]: that's actually interesting.
The .strcat() is probably legacy C, which got replaced by new mechanics of C++
That += but not + is calling .append() is interesting, since both could easily be set up to call append

File: src/view/terminalview.h (hunk 3 above)
Comment (xeruf) [reply src/view/terminalview.h]: but we need access for unit tests, otherwise we can't properly do TDD :/

Comment (anarchuser) [reply src/view/terminalview.h]: then make it protected and subclass it

Comment (anarchuser) [reply src/view/terminalview.h]: I don't know what the general approach is, but I'd value proper data encapsulation and privacy protection over unit tests

File: CMakeLists.txt [hunk 6]
```diff
@@ -9,48 +9,67 @@ eve
...
+<<<<<<< HEAD
 src/model/object/entity/individual/fred/fred.cpp
 src/model/object/entity/individual/fred/fred.h
+=======
 src/model/object/entity/entity.cpp
-src/model/object/object.cpp
+>>>>>>> feat(view): add terminalview
+=======
+src/model/object/entity/entity.cpp
+>>>>>>> a3a4422efdc38b07a3e47a67ab26c333115509db
```
Comment (xeruf) [inline CMakeLists.txt]: unresolved conflict?

File: CMakeLists.txt [hunk 7]
```diff
@@ -9,48 +9,67 @@ eve
...
```
Comment (xeruf) [inline CMakeLists.txt]: ^

File: src/model/world/world.h [hunk 8]
```diff
@@ -19,7 +19,7 @@ class World {
     int energy = 0;
 
     std::vector<Food> food;
-    std::vector<Entity> entities;
+    std::vector<Individual> individuals;
```
Comment (xeruf) [inline src/model/world/world.h]: why is this here?

File: src/view/terminalview.cpp [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    }
+    return formattedString;
+};
+
+std::string TerminalView::stringifyFood(Food food) {
+    int xPos;
+    int yPos;
+    int value;
+
+    xPos = food.getPosition().x;
```
Comment (xeruf) [inline src/view/terminalview.cpp]: I'd suggest merging declaration & assignment

File: src/view/terminalview.h (hunk 5 above)
Comment (anarchuser) [reply src/view/terminalview.h]: I correct myself:
I think the classes can't be const
if they are const you might not be able to even call methods

**Review by xeruf — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 6 above)
Comment (anarchuser) [reply CMakeLists.txt]: I think we can safely ignore any changes in this file.
Simply call `./sort_cmklist.sh > CMakeLists.txt`

Comment (anarchuser) [reply CMakeLists.txt]: just resolve it somehow

File: src/view/terminalview.h (hunk 3 above)
Comment (AtomToast) [reply src/view/terminalview.h]: I read that the general process seems to be to outsource such functions into a seperate class to keep things testable

File: src/model/world/world.h (hunk 8 above)
Comment (AtomToast) [reply src/model/world/world.h]: I don't know, I don't work on that file

File: src/view/terminalview.cpp (hunk 9 above)
Comment (AtomToast) [reply src/view/terminalview.cpp]: If I remember correctly the linter actually recommended this but I can revert it back to having it in one line

File: src/view/terminalview.h (hunk 5 above)
Comment (AtomToast) [reply src/view/terminalview.h]: https://stackoverflow.com/questions/7626116/how-to-pass-object-to-function-in-c
So basically if I don't want to change the original object do nothing and if I don't intend to or it doesn't matter use &?

File: src/view/terminalview.cpp (hunk 9 above)
Comment (xeruf) [reply src/view/terminalview.cpp]: please do not resolve such conversations! The commenter should resolve them.

I am wondering why the linter would recommend that.

File: test/view/terminalview.cpp [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        "X:   43 | Y:  978 | Value:  20\n"
+        "—————————————————————————————————————————\n"
+        "Entities:\n"
+        "—————————————————————————————————————————\n"
+        "ID:  23 | X:  123 | Y:  321 | Energy: 100\n"
+        "ID:  24 | X:  138 | Y:  153 | Energy:  70\n"
+        "ID:  69 | X:  420 | Y:  420 | Energy:   5\n";
+
+    World world = World(1000, 1000, 40);
+    SECTION(""){}
```
Comment (xeruf) [inline test/view/terminalview.cpp]: wats dis

**Review by xeruf — CHANGES_REQUESTED** [review]

File: src/model/world/world.h (hunk 8 above)
Comment (anarchuser) [reply src/model/world/world.h]: that's why this shouldn't be in this pull request.
(besides the fact that this is already in the master)

File: test/view/terminalview.cpp (hunk 10 above)
Comment (AtomToast) [reply test/view/terminalview.cpp]: outdated placeholder

File: CMakeLists.txt [hunk 11]
```diff
@@ -1,6 +1,7 @@
 cmake_minimum_required(VERSION 3.14)
 set(CMAKE_CXX_STANDARD 17)
 
+
```
Comment (xeruf) [inline CMakeLists.txt]: bitte entfernen

File: src/view/terminalview.cpp [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+
+    std::vector<Individual *> individuals = world.getIndividuals();
+    for (int i = 0; i < individuals.size(); ++i) {
+        outputString += stringifyIndividual(individuals[i]);
+    }
+
+    return outputString;
+};
+
+std::string TerminalView::formatNumberSpacing(int number, int maxLen) {
```
Comment (xeruf) [inline src/view/terminalview.cpp]: das könnte eine generelle Methode sein mit dem Namen einfach nur `padLeft` oder so, und statt "number" kann es eigentlich einen beliebigen parameter nehmen?

File: src/view/terminalview.cpp [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+    return outputString;
+};
+
+std::string TerminalView::stringifyIndividual(Individual * individual) {
+    int id = individual->getID();
+    int xPos = individual->getPosition().x;
+    int yPos = individual->getPosition().y;
+    int energy = individual->getEnergy();
+
+    std::string formattedID = formatNumberSpacing(id, 3);
```
Comment (xeruf) [inline src/view/terminalview.cpp]: das kannst du eigentlich alles inlinen

File: CMakeLists.txt (hunk 11 above)
Comment (AtomToast) [reply CMakeLists.txt]: I merged that from master but I can still remove it

File: src/view/terminalview.cpp (hunk 12 above)
Comment (AtomToast) [reply src/view/terminalview.cpp]: No, it only takes numbers. It doesn't make sense to overload the method since there is nothing else other than numbers being padded. I could change the name of the function but that seems like kind of a trivial change

Comment (AtomToast) [reply src/view/terminalview.cpp]: I am considering changing the logic of the function though to not use a while loop but rather just add the appropriate number of spaces by utilising the string constructor

File: src/view/terminalview.cpp (hunk 13 above)
Comment (AtomToast) [reply src/view/terminalview.cpp]: What do you mean? I tried to seperate them into relevant blocks to keep it easily readable if that's what you mean. Otherwise could you maybe suggest a change?

File: src/main.cpp [hunk 14]
```diff
@@ -4,5 +4,27 @@ int main(int argc, char * argv[]) {
...
+
+    world.addIndividual(fred1);
+    world.addIndividual(fred2);
+    world.addIndividual(fred3);
+
+    world.addFood(food1);
+    world.addFood(food2);
+    world.addFood(food3);
+
+    terminalView.render(world);
```
Comment (anarchuser) [inline src/main.cpp]: why is this in the main.cpp file??

Comment (anarchuser) [reply src/main.cpp]: the controller initialises a world with individuals and food.
all you need to do is incorporate it in the controller.
otherwise, just put it in test files.

But, please, refrain from putting debug stuff like this in our main prod code

File: src/main.h [hunk 15]
```diff
@@ -2,6 +2,8 @@
 #define EVE_MAIN_H
 
 #include "controller/controller.h"
+#include "model/world/world.h"
+#include "view/terminalview.h"
```
Comment (anarchuser) [inline src/main.h]: (see main.cpp)
this belongs into controller.h

File: src/view/terminalview.cpp [hunk 16]
```diff
@@ -51,6 +57,21 @@ std::string TerminalView::stringifyFood(Food & food) {
     return outputString;
 };
 
-std::string TerminalView::stringifyEntity(Entity & entity) {
+std::string TerminalView::stringifyIndividual(Individual * individual) {
```
Comment (anarchuser) [inline src/view/terminalview.cpp]: instead of creating stringifyFood and stringifyIndividual you could also keep stringifyEntity and overload it with Food or Individual
(or, for even more generic naming, call it stringify only)

File: src/main.cpp (hunk 14 above)
Comment (AtomToast) [reply src/main.cpp]: Janek told me to. I just wanted to put it in there to test stuff but he said I should also commit it.
All I heard from him is that the controller doesn't do anything

File: src/view/terminalview.cpp (hunk 16 above)
Comment (AtomToast) [reply src/view/terminalview.cpp]: I considered that since the functions are quite similar. But I might have forgotten even though I wrote a reminder comment above them

File: src/main.cpp (hunk 14 above)
Comment (anarchuser) [reply src/main.cpp]: Janek's not responsible for the controller.
If you want to test it in the main file, please put all this in the controller.

The controller initialises the world; you can just change the implementation of `.init()` to however you want to initialise the world.
Also, in `.simulate()`, call the `.render()` function, because that's exactly the place where things will be rendered in the future anyways.

Everything which currently is in `.init()` and `.simulate()` is only there for testing, you can just replace it by how you need it

File: src/view/terminalview.cpp (hunk 16 above)
Comment (AtomToast) [reply src/view/terminalview.cpp]: I think why I didn't do it was because the individual also gives the ID. But still I guess I could just do it by overloading

Comment (anarchuser) [reply src/view/terminalview.cpp]: that's no reason to not overload it.
Overloading means you write different functions depending on the input, so in stringify(Individual) you can print everything plus ID, and in stringify(Food) you only print position and food

Comment (anarchuser) [reply src/view/terminalview.cpp]: Tbh, you could just have one function stringify(Object) and read out what specific type it is - saving you even more duplicate code XD

Comment (anarchuser) [reply src/view/terminalview.cpp]: I'm looking forward to see solve this without writing much duplicate code

File: CMakeLists.txt (hunk 11 above)
Comment (AtomToast) [reply CMakeLists.txt]: and it breaks the compilation, as expected

File: src/view/terminalview.cpp (hunk 12 above)
Comment (anarchuser) [reply src/view/terminalview.cpp]: I mean
you could change it to a very generic function taking one arbitrary parameter, the supposed string length, and whether to align it le/mi/ri.
Mayber later we add other things where we need it to work with other things than numbers.

Comment (anarchuser) [reply src/view/terminalview.cpp]: also, please be aware that "Number" mathematically refers to any kind of number; not only "natural numbers" like ints, but also floating point, irrational and complex numbers.
that's not really an issue, it just might confuse people that it only ever takes an int, and not, e.g., a float

File: src/main.cpp (hunk 14 above)
Comment (xeruf) [reply src/main.cpp]: yeah listen to Aaron, I am unaware of the current state here :)

File: CMakeLists.txt (hunk 11 above)
Comment (AtomToast) [reply CMakeLists.txt]: ok removed

File: src/view/terminalview.cpp (hunk 12 above)
Comment (AtomToast) [reply src/view/terminalview.cpp]: I don't think that's really needed right now so I shouldn't make it unnecessarily complex

File: src/view/terminalview.cpp (hunk 13 above)
Comment (AtomToast) [reply src/view/terminalview.cpp]: done

Comment (AtomToast) [reply src/view/terminalview.cpp]: commit 46ca073238829b9b1c997422f648d658927a3ac5

File: CMakeLists.txt (hunk 11 above)
Comment (AtomToast) [reply CMakeLists.txt]: commit 8223effb6c4b6f06076d2056d9244e4af7d498e8

File: src/view/terminalview.cpp (hunk 16 above)
Comment (AtomToast) [reply src/view/terminalview.cpp]: commit 46ca073238829b9b1c997422f648d658927a3ac5

**Comment by AtomToast** [comment]
~~- [ ] add ncurses~~
going to do this on another branch

File: src/view/terminalview.cpp (hunk 12 above)
Comment (AtomToast) [reply src/view/terminalview.cpp]: commit 798385a8b2e10d06f0d4c2917cf2279194084543

Comment (AtomToast) [reply src/view/terminalview.cpp]: Hm, maybe I could avoid the len variable by inlining it, thoughts?

File: src/main.cpp (hunk 14 above)
Comment (AtomToast) [reply src/main.cpp]: commit 6ff2b4e

File: src/main.h (hunk 15 above)
Comment (AtomToast) [reply src/main.h]: commit 6ff2b4e

File: src/view/terminalview.cpp (hunk 12 above)
Comment (AtomToast) [reply src/view/terminalview.cpp]: commit fad6b2237c5fa881ec634659fec7c9f5b55d14b3

**Review by anarchuser — APPROVED** [review]
Looks good. I think all the things pointed out were improved properly

**Comment by xeruf** [comment]
@anarchuser please don't forget to resolve your conversations so you leave this PR clean ;)

**Review by xeruf — APPROVED** [review]
